lib/edge_impulse/ei_ncs.cpp

```diff
@@ -0,0 +1,399 @@
+
+
+static size_t buf_calc_max_processing_move(const struct data_buffer *b)
+{
+	const float *buf_start = b->buf;
+	const float *buf_end = buf_start + ARRAY_SIZE(b->buf);
+
+	if (b->append_head >= b->processing_head) {
+		return b->append_head - b->processing_head;
+	}
+
+	return (buf_end - b->processing_head) + (b->append_head - buf_start);
+}
+
+static size_t buf_calc_free_space(const struct data_buffer *b)
+{
+	const float *buf_start = b->buf;
+	const float *buf_end = buf_start + ARRAY_SIZE(b->buf);
+
+	if (b->wait_data_size > 0) {
+		return b->wait_data_size + ARRAY_SIZE(b->buf) -
+		       INPUT_WINDOW_SIZE - 1;
+	} else if (b->processing_head > b->append_head) {
+		return b->processing_head - b->append_head - 1;
+	}
+
+	return (buf_end - b->append_head) +
+	       (b->processing_head - buf_start) - 1;
+}
+
+static void buf_processing_end(struct data_buffer *b)
+{
+	k_spinlock_key_t key = k_spin_lock(&b->lock);
+
+	__ASSERT_NO_MSG(b->state == STATE_PROCESSING);
+	b->state = STATE_READY;
+
+	k_spin_unlock(&b->lock, key);
+}
+
+static int buf_cleanup(struct data_buffer *b)
+{
+	int err = 0;
+
+	k_spinlock_key_t key = k_spin_lock(&b->lock);
+
+	if (b->state == STATE_PROCESSING) {
+		err = -EBUSY;
+	} else {
+		b->processing_head = b->buf;
+		b->append_head = b->buf;
+		b->wait_data_size = 0;
+		b->state = STATE_READY;
+	}
+
+	k_spin_unlock(&b->lock, key);
+
+	return err;
+}
+
+static int buf_append(struct data_buffer *b, const float *data, size_t len,
+		      bool *process_buf)
+{
+	float *buf_start = b->buf;
+	float *buf_end = buf_start + ARRAY_SIZE(b->buf);
+
+	*process_buf = false;
+
+	k_spinlock_key_t key = k_spin_lock(&b->lock);
+
+	if (buf_calc_free_space(b) < len) {
+		k_spin_unlock(&b->lock, key);
+		return -ENOMEM;
+	}
+
+	float *cur_head = b->append_head;
+	float *new_head = b->append_head + len;
+	bool looped = false;
+
+	if (b->wait_data_size > 0) {
+		if (b->wait_data_size > len) {
+			b->wait_data_size -= len;
+		} else {
+			b->wait_data_size = 0;
+			*process_buf = true;
+		}
+	}
+
+	if (new_head >= buf_end) {
+		new_head -= ARRAY_SIZE(b->buf);
+		looped = true;
+	}
+
+	b->append_head = new_head;
+
+	k_spin_unlock(&b->lock, key);
+
+	if (looped) {
+		size_t buf_left = buf_end - cur_head;
+
+		memcpy(cur_head, data, buf_left * sizeof(b->buf[0]));
+		memcpy(buf_start, data + buf_left,
+		       (len - buf_left) * sizeof(b->buf[0]));
+	} else {
+		memcpy(cur_head, data, len * sizeof(b->buf[0]));
+	}
+
+	return 0;
+}
+
+static void buf_get(const struct data_buffer *b, float *b_res, size_t offset,
+		    size_t len)
+{
+	__ASSERT_NO_MSG((offset + len) <= INPUT_WINDOW_SIZE);
+
+	const float *buf_start = b->buf;
+	const float *buf_end = buf_start + ARRAY_SIZE(b->buf);
+
+	/* Processing head cannot change while processing is done. */
+	__ASSERT_NO_MSG(b->state == STATE_PROCESSING);
+
+	const float *read_start = b->processing_head + offset;
+	const float *read_end = read_start + len;
```

Comment:
this is undefined behavior in c
```
— Addition or subtraction of a pointer into, or just beyond, an array object and an
integer type produces a result that does not point into, or just beyond, the same array
object (6.5.6).
```

You cannot add anything to the pointer that will make the pointer go beyond `&array[last_el+1]`. The dangerous part in the code here is used to control the logic. Optimizer may assume that len will never go beyond array (since this is UB it can assume anything about it) and will then simply get rid of case in line 177.

For this reason it is safer to operate on element indexes not pointers.

Reply:
I will rework this part to use indexes instead of pointers.